src/library/scala/collection/immutable/ListSet.scala

```diff
@@ -110,14 +110,37 @@ sealed class ListSet[A] extends AbstractSet[A]
             var toAdd = ls
             var result: ListSet[A] = this
 
+            var pending: Array[A] = null
+            var pending0, pending1, pending2, pending3: A = null.asInstanceOf[A]
+            var pendingCount = 0
             while (toAdd ne lsTail) {
               val elem = toAdd.elem
               if (!containsLimited(result, elem, lsTail)) {
-                val r = result
-                result = new r.Node(elem)
+                pendingCount match {
+                  case 0 => pending0 = elem
+                  case 1 => pending1 = elem
+                  case 2 => pending2 = elem
+                  case 3 => pending3 = elem
+                  case _ =>
+                    if (pending eq null) pending = new Array[AnyRef](lsSize - remaining - 4).asInstanceOf[Array[A]]
```

Comment:
the size for this `Array` is incorrect—it will be negative when `lsSize < thisSize` and `ls` has more than four elements that are not in `this`.

```scala
scala> ListSet(1, 2, 3, 4, 5, 6, 7, 8) ++ ListSet(9, 10, 11, 12, 13)
java.lang.NegativeArraySizeException
  at scala.collection.immutable.ListSet.$plus$plus(ListSet.scala:125)
  ... 28 elided
```

once fixed, you should probably add that as a junit test as well

--------

H/T @dwijnand for making this super easy to test using `scala-runners`

Reply:
>  for making this super easy to test using scala-runners

(for those who don't know, with https://github.com/dwijnand/scala-runners you can `scala --scala-pr 9441`)

Reply:
fixed and added test

Reply:
fixed but broken again D:

```scala
ListSet(1, 2, 3, 4, 5, 6, 7, 8) ++ ListSet(9, 10, 11, 12, 13)
java.lang.ArrayIndexOutOfBoundsException: 0
  at scala.runtime.ScalaRunTime$.array_update(ScalaRunTime.scala:74)
  at scala.collection.immutable.ListSet.$plus$plus(ListSet.scala:137)
  ... 28 elided
```